src/marshal.c

```diff
@@ -227,13 +237,98 @@ int format_packet(bot_t *bot, void *_packet_data, void *_packet_raw)
     }
 
     varlen = varint32_encode(packet_raw - save, varint, 5);
-    if(packet_raw - save + varlen > len)
+    if(packet_raw - save + varlen > len) {
```

Comment:
Yeah I'm really sick of this boilerplate bounds checking code every time we add to the buffer. I think we should consider using a buffer struct, possibly even the one that cNBT has (there's already code in mcc that uses it, see line 279 and 322). That saves us passing the length around and we can modify push() to use it.